src/bit_machine/limits.rs

```diff
@@ -0,0 +1,120 @@
+///
+/// This roughly corresponds to the maximum size of any type used by a
+/// program, in bits. In a blockchain context this should be limited by
+/// the transaction's weight budget.
+///
+/// The limit here is an absolute limit enforced by the library to avoid
+/// unbounded allocations.
+// This value must be less than usize::MAX / 2 to avoid panics in this module.
+const MAX_CELLS: usize = 2 * 1024 * 1024 * 1024 - 1;
+
+/// The maximum number of frames needed by the bit machine.
+///
+/// This roughly corresponds to the maximum depth of nested `comp` and
+/// `disconnect` combinators. In a blockchain context this should be
+/// limited by the transaction's weight budget.
+///
+/// The limit here is an absolute limit enforced by the library to avoid
+/// unbounded allocations.
+// This value must be less than usize::MAX / 2 to avoid panics in this module.
+const MAX_FRAMES: usize = 1024 * 1024;
+
+#[non_exhaustive]
+#[derive(Clone, Debug)]
+pub enum LimitError {
+    MaxCellsExceeded {
+        /// The number of cells needed by the program.
+        got: usize,
+        /// The maximum allowed number of cells.
+        max: usize,
+        /// A description of which cell count exceeded the limit.
+        bound: &'static str,
+    },
+    MaxFramesExceeded {
+        /// The number of frames needed by the program.
+        got: usize,
+        /// The maximum allowed number of frames.
+        max: usize,
+        /// A description of which frame count exceeded the limit.
+        bound: &'static str,
+    },
+}
+
+impl fmt::Display for LimitError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let (limit, got, max, bound) = match self {
+            LimitError::MaxCellsExceeded { got, max, bound } => ("cells", got, max, bound),
+            LimitError::MaxFramesExceeded { got, max, bound } => ("frames", got, max, bound),
+        };
+        write!(
+            f,
+            "maximum number of {} exceeded (needed {}, maximum {}) (bound: {})",
+            limit, got, max, bound,
+        )
+    }
+}
+impl error::Error for LimitError {}
+
+impl LimitError {
+    fn check_max_cells(got: usize, bound: &'static str) -> Result<(), Self> {
+        if got > MAX_CELLS {
+            Err(Self::MaxCellsExceeded {
+                got,
+                max: MAX_CELLS,
+                bound,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    fn check_max_frames(got: usize, bound: &'static str) -> Result<(), Self> {
+        if got > MAX_FRAMES {
+            Err(Self::MaxFramesExceeded {
+                got,
+                max: MAX_CELLS,
+                bound,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Helper function to check every value and sum for being within bounds.
+    pub(super) fn check_program(
+        source_ty_width: usize,
+        target_ty_width: usize,
+        bounds: &crate::NodeBounds,
+    ) -> Result<(), Self> {
```

Comment:
2110376b0556fe2bbc1e2a5e79f04ed68f446ba3: Is there a reason why the argument list is not simply `program: &RedeemNode<J>` ?

Reply:
Not really, no. IIRC I didn't want to type `<J: Jet>`. I don't remember why I thought that typing `crate::NodeBounds` would be any less annoying.

Let me know if you want to change it.

Reply:
Let's change the parameter to `&RedeemNode<J>`, since that is the only node type that knows its bounds.